crypto/bp/bp_map.c

```diff
@@ -0,0 +1,648 @@
+             */
+            if (!G2_ELEM_copy(_q[m], q[i]))
+                goto err;
+            if (!G2_ELEM_make_affine(group, _q[m], ctx))
+                goto err;
+            if (!FP2_copy(x[m], _q[m]->X))
+                goto err;
+            if (!FP2_copy(y[m], _q[m]->Y))
+                goto err;
+            if (!FP2_copy(xq[m], _q[m]->X))
+                goto err;
+            if (!FP2_copy(yq[m], _q[m]->Y))
+                goto err;
+            if (!FP2_copy(zq[m], _q[m]->Z))
+                goto err;
+            m++;
+        }
+
+    if (m == 0)
+        return GT_ELEM_set_to_unity(group, r);
+
+    /*
+     * Initialize line function.
+     */
+    if ((l = FP12_new()) == NULL)
+        goto err;
+    if (!FP12_zero(l))
+        goto err;
+
+    /*
+     * Initialize Miller variables.
+     */
+    if (!FP12_zero(r->f))
+        goto err;
+    if (!BN_copy(r->f->f[0]->f[0]->f[0], group->one))
+        goto err;
+    if (!BN_copy(miller, group->param))
+        goto err;
+    if (!BN_mul_word(miller, 6))
+        goto err;
+    if (!BN_add_word(miller, 2))
+        goto err;
+
+    /*
+     * Compute Miller loops simultaneously.
+     */
+    for (i = BN_num_bits(miller) - 1; i > 0;) {
+        i--;
+        if (!FP12_sqr(group, r->f, r->f, ctx))
+            goto err;
+        for (j = 0; j < m; j++) {
+            if (!GT_miller_double
+                (group, l, xq[j], yq[j], zq[j], xq[j], yq[j], zq[j], s[j],
+                 t[j], ctx))
+                goto err;
+            if (!FP12_mul_sparse(group, r->f, r->f, l, ctx))
+                goto err;
+        }
+        if (BN_is_bit_set(miller, i)) {
+            for (j = 0; j < m; j++) {
+                if (!GT_miller_add
+                    (group, l, xq[j], yq[j], zq[j], x[j], y[j], xp[j], yp[j],
+                     ctx))
+                    goto err;
+                if (!FP12_mul_sparse(group, r->f, r->f, l, ctx))
+                    goto err;
+            }
+        }
+    }
+
+    if (BN_is_negative(group->param) && !FP12_conj(group, r->f, r->f))
+        goto err;
+
+    for (i = 0; i < m; i++) {
+        if (BN_is_negative(group->param) && !FP2_neg(group, yq[i], yq[i]))
+            goto err;
+        if (!GT_miller_final
+            (group, r->f, xq[i], yq[i], zq[i], x[i], y[i], xp[i], yp[i],
+             ctx)) {
+            goto err;
+        }
+    }
+    if (!GT_final_exp(group, r->f, r->f, ctx))
+        goto err;
+
+    ret = 1;
+
+ err:
+    BN_CTX_end(ctx);
+    BN_CTX_free(new_ctx);
+    for (i = 0; i < num; i++) {
+        FP2_free(x[i]);
+        FP2_free(y[i]);
+        FP2_free(xq[i]);
+        FP2_free(yq[i]);
+        FP2_free(zq[i]);
+        G2_ELEM_free(_q[i]);
+    }
+    FP12_free(l);
+    return ret;
+}
+
+int GT_ELEM_pairing(const BP_GROUP *group, GT_ELEM *r, const G1_ELEM *p,
+                    const G2_ELEM *q, BN_CTX *ctx)
+{
+    const G1_ELEM *_p[1];
```

Comment:
nit: `ps` and `qs` instead of `_p` and `_q`? Underscore variables aren't our usual style.